rcl_lifecycle/src/default_state_machine.c

```diff
@@ -690,10 +690,27 @@ rcl_lifecycle_init_default_state_machine(
 
   return ret;
 
-fail:
+// Semicolon handles the "a label can only be part of a statement..." error
+fail:;
+  // If rcl_lifecycle_transition_map_fini() fails, it will clobber the error string here, twice
+  // Here, we concatenate the error strings if that happens
+  const char * current_error = (rcl_error_is_set()) ? rcl_get_error_string().str : "";
+  rcl_reset_error();
+
   if (rcl_lifecycle_transition_map_fini(&state_machine->transition_map, allocator) != RCL_RET_OK) {
-    RCL_SET_ERROR_MSG("could not free lifecycle transition map. Leaking memory!\n");
+    const char * fini_error = (rcl_error_is_set()) ? rcl_get_error_string().str : "";
+    rcl_reset_error();
+
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Freeing transition map failed while handling a previous error. Leaking memory!"
+      "\nOriginal error:\n\t%s\nError encountered in rcl_lifecycle_transition_map_fini():\n\t%s\n",
+      current_error, fini_error);
+  } else if (strcmp(current_error, "") != 0) {
+    RCL_SET_ERROR_MSG(current_error);
```

Comment:
Sorry to iterate again over this, but I think that `else if` branch is not needed. If I see this correctly, we'll fetch the error message before just to reset it again in this branch. 
Why not just retrieving the existing error message within the first `if` branch? 

something like this:
```c++
if (rcl_lifecycle_transition_map_fini(&state_machine->transition_map, allocator) != RCL_RET_OK) {
  const char * fini_error = "";
  if (rcl_error_is_set()) {
     fini_error = rcl_get_error_string().str;
     rcl_reset_error();
  }
  RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
  "Freeing transition map failed while handling a previous error. Leaking memory!"
  "\nOriginal error:\n\t%s\nError encountered in rcl_lifecycle_transition_map_fini():\n\t%s\n",
  current_error, fini_error);
}

if (!rcl_error_is_set()) {
  RCL_SET_ERROR_MSG("Unspecified error ..");
}

```

Reply:
The problem originally was that the error message was already set. But if `rcl_lifecycle_transition_map_fini` failed, it would set its own message, and then the message would get set a third time inside this first if loop. So the goal was to keep track of `current_message`, and append fini_error if it existed and throw a combined error that was hopefully more helpful than either of them individually.

However, if `rcl_lifecycle_transition_map_fini` succeeds, we still want the user to know what caused the code to reach `fail:` in the first place, which is in `current_message` , but only if rcl_error_is_set() was true in line 697.

Originally when I submitted this, I just relied on rcutils to log the transition of error messages, since it warns if a new error message is set while a previous one had already existed. However, since I'm pretty new to the ROS 2 C codebase, I'm really not confident about any particular approach here :man_shrugging: